### ISSUE TITLE
Rewrite React-Redux docs completely

### ISSUE THREAD

**Issue by markerikson** [reduxjs/react-redux#1001]
The current React-Redux docs aren't very helpful, for a lot of reasons:

- They're just a Markdown file in the repo, instead of being published in HTML form
- It feels very "dense" and jargon-heavy
- It doesn't clearly lay out the various options for `connect` and ways that you can use it
- It's primarily in API reference form, rather than a "how to"

We do also have the "Using React with Redux" page over in the main Redux docs, which is a bit more of a "getting started" page, but it's also not great.

So, I would really like to completely rewrite our React-Redux docs from scratch.

As a rough outline, what I'd like to see is:

- Quick Start (copy-pasteable examples that show how to add `<Provider>` and do a basic call to `connect()`
- How It Works (an explanation of how `<Provider>` makes the store accessible, and roughly what `connect` does internally to subscribe and extract data)
- Advanced Techniques / Recipes (examples of things like using the "factory function" syntax for per-component selector memoization)
- API Reference (something similar to what we have now)

I do already have a Gitbook setup configured for this, same as the main Redux docs.  I should be able to get the `react-redux.js.org` domain name for this.

**Comment by markerikson** [comment]
Awesome, thanks - really appreciate it!  

I'd previously discussed this with @jsonnull .  One note from a discussion:

> Since I was envisioning a breakdown into "High level concepts" ... "Recipes" ... "API Docs", I thought about making a PR which simply pulls apart the current docs into those categories.

Might be a place to start.

**Comment by timdorr** [comment]
Since Redux + React is such a common combo, I'd rather keep it all under the redux.js.org site. There's already a ton of React code in those docs anyways. And it also has the side effect of being more discoverable that way (one single search bar for both libraries). 

We can point everything here to that page (or pages!). We can also consolidate the API reference, so there's one place for all the API docs.

**Comment by markerikson** [comment]
@wgao19 : Hiya! Thanks for looking into this.  A couple quick thoughts:

- I did set up the initial Gitbook integration a while back.  It's viewable at https://redux.gitbook.io/react-redux/ .  I can see us going either way on whether it eventually becomes part of the main Redux docs page at https://redux.js.org , or whether it stays as a separate page (in which case I would want to get the `https://react-redux.js.org` domain name).  In any case, we can focus on the content part for now, and worry about the publishing aspect later.  
- I've got links to some really good "intro to React and Redux" tutorials in the [Learning Resources#Using Redux with React](https://redux.js.org/introduction/learningresources#using-redux-with-react) page in the Redux docs.  I would say we should try to use some of those as sources of inspiration for the "quick start" aspect.  And yes, we can absolutely borrow from my slides for the "how it works" part.
- I talked about the "factory function" approach in my post [Idiomatic Redux: Using Reselect Selectors for Encapsulation and Performance](https://blog.isquaredsoftware.com/2017/12/idiomatic-redux-using-reselect-selectors/).  In a new docs structure, we'd probably want to talk about the basics of using selectors in some kind of "Writing `mapState` Functions" or "Using `mapState`" recipes page, and then save the "factory function" aspect for an "Advanced Performance Optimizations" page.

@ConnorBryan , have you had a chance to look into any of this yet?  Hopefully the two of you can collaborate on this.

**Comment by wgao19** [comment]
Thank you @markerikson that is a lot more learning resources to look into. I will read and watch them in the following few days and start drafting stuff.
@ConnorBryan It'd be great if we can collaborate on this! Though I'm not very experienced with open source contribution, perhaps I should pull up a PR that we can work on together?

**Comment by ConnorBryan** [comment]
@wgao19 I would be honored to work on this with you; are you on Discord at all? That's my primary place to communicate in real time; I think that might be the best way to hash this out.

**Comment by wgao19** [comment]
hey @ConnorBryan yes I am on discord @wgao19#0252

**Comment by ConnorBryan** [comment]
@wgao19 I'm having trouble locating you -- can you add `connor#5456` ?

**Comment by markerikson** [comment]
Per discussion, we're setting up a Dropbox Paper doc to throw together an outline, same as we did for the main Redux docs:

https://paper.dropbox.com/doc/React-Redux-docs-draft-outline--AL9W1hkb6TLJ5se2~6G4awnSAQ-z5BoKAQVMVCo1SxGUhqbH

**Comment by maecapozzi** [comment]
I would also love to participate in the rewrite of the docs, if you'll have me. I have a bit of experience with redux, and a lot of experience with React. I also have written quite a few popular tutorials on React as well as webpack, so I might be helpful there. Is there still room for contributions @markerikson?

**Comment by markerikson** [comment]
@maecapozzi : sent a DM over on Discord/Reactiflux, if you can go check that.  Thanks!

**Comment by Canopix** [comment]
Are you thinking about the translation in Spanish?  May I help with the translation.

**Comment by timdorr** [comment]
Unfortunately, there isn't multi-lingual support built-in just yet: https://docs.gitbook.com/v2-changes/important-differences#multi-language-books

**Comment by Canopix** [comment]
Ohh, that's disappointing... Maybe with markdown we could make the docs on Spanish :(

**Comment by markerikson** [comment]
Okay, just merged in the new "Getting Started" page at #1017 .  Looks great!

**Comment by wgao19** [comment]
Thank you @markerikson !
The Gitbook generated page does look a bit off. It seems that the generated table of content on the right is broken by the code as well (._.) Since we also have the CodeSandbox links, perhaps we can remove the code wrapped with `<details>` for now? 
Also I notice that the title is in CamelCase. How should we fix that?

**Comment by markerikson** [comment]
Yeah, let's delete the hidden code blocks for now.  Other than that, don't worry too much about the Gitbook formatting - I'll need to figure out how to fiddle with the setup appropriately.

**Comment by markerikson** [comment]
For the "dispatching actions" page, Dan already has a good writeup on SO that we can base it on: https://stackoverflow.com/questions/34458261/how-to-get-simple-dispatch-from-this-props-using-connect-w-redux/34458710#34458710

**Comment by markerikson** [comment]
And, Dave Ceddia just posted an excellent [You Might Not Need `mapDispatchToProps`](https://daveceddia.com/redux-mapdispatchtoprops-object-form/) post, inspired by my talk at ReactBoston.  Great fodder for "dispatching actions".

**Comment by markerikson** [comment]
We now have the React-Redux docs being published at https://react-redux.js.org !  Built with Docusaurus, hosted on Netlify, and we've got deploy previews turned on for PRs.  I'm very excited about this!

**Comment by sw-yx** [comment]
:) :) :) mark please ping me if i can help with anything on the netlify side. super glad and hope the deploy previews help with the rest of the ongoing docs rewrite

**Comment by markerikson** [comment]
Yep, loving it so far!

**Comment by zsid** [comment]
Hi everyone. How can I help with writing the docs? What will be best to start with?

**Comment by markerikson** [comment]
@zsid : Hi!  You can see a list in the previous comments of some of the other pages we'd like to add.

Perhaps a good place to start would be to start adapting my post [Idiomatic Redux: Using Reselect Selectors for Performance and Encapsulation](https://blog.isquaredsoftware.com/2017/12/idiomatic-redux-using-reselect-selectors/) into a docs page.  If you'd like to work one of the other possible pages as well, that would also be great.

**Comment by zsid** [comment]
I will start with your suggestion @markerikson and work on selectors 😁Thanks.

**Comment by markerikson** [comment]
We should _totally_ add a "Why Should You Use React-Redux?" section to the docs, in the intro section.  Base it on my workshop slides.

**Comment by wgao19** [comment]
Hi everyone,
I'm working on the "Why React-Redux" article now.
I have a working draft [here](https://paper.dropbox.com/doc/Why-React-Redux--AQwiY_XlVi7ndBIpx2jCHkP8AQ-Js94Eh9ilcgwfzfus23TD), welcoming suggestions :)

**Comment by codedavinci** [comment]
Hello Y'all, 

@markerikson quick suggestion for the intro, I noticed we have "Why use react-redux", I think it'd be really import to have a section describing "When to use `react-redux` " mentioning some of use cases that you could use for instance`React.createContext` to float state around that application, `redux.getState()` or simply raw `state`.  Nowadays there several ways of managing `state` in a raw fashion with `react`, but we still face some problems to flowy manipulate the **methods (AKA Actions)**  that update this abundant `state` , and that's `react-redux` - `connect`  comes to play !  what do you think ?

Introduction
- [ ]   Quick Start
- [ ]   Basic Tutorial
- [ ]   Why Use React-Redux?
- [ ]   When Use React-Redux?

**Comment by markerikson** [comment]
Hmm.  To some extent, it's not really a question of "When to Use React-Redux", but rather a question of "when should you use Redux in a React app, vs other options".

I'm somewhat less concerned about that topic as a whole, but I could see it being a reasonable thing to include.  If you'd like to put something together, go ahead.

**Comment by codedavinci** [comment]
@markerikson, I am totally with you on that !
but It's also important to extend the knowledge that we **can** use Redux without the `connect`, with the api `Redux.getState()` + methods on demand through `Redux.subscribe` to update the `store`. 
Actually, the majority of newcomers think that `React-Redux` is a requirement in order to utilize  `Redux` which's not the case. 

I think the place where I am trying to get is to demonstrate that the `React-redux` is optional in order to use `Redux`

Goal:  draw a picture of a "life" without `react-redux` (`Provider` and `connect` ), using just `redux`. 
And another picture using `redux` + `react-redux`,  showcasing all the advantages of using (`Provider` and `connect`).  

From an educational prospective, I believe that having a section that explain the topics above prior is the key to highlight the "Why use react-redux" section.  It's  extremely important for newcomers to see the difference, so they understand why they are choosing to adopt `react-redux` consciously, not just for tutorial /tooling  habits. 

I am gonna draft up a Why Use React-Redux page to follow the thread and I'll send a WIP-PR for you to take a look on the format. 

Perhaps that can be implicitly mentioned at the top of the section of Why to use.. 

What do you think?

Thanks for the input !

**Comment by markerikson** [comment]
Some updates today:

I bumped us up to Docusaurus 1.6.0, modified the config to remove `/docs/` from the URLs, and changed the IDs for the `mapState` and `mapDispatch` pages to produce shorter URLs.  I also added a Netlify redirects file to catch any existing incoming links to the old URLs.  See #1096 .

After looking at the couple proposals for a "Why Use React-Redux?" page, I wound up writing a new one in #1097 and merged it.  It's viewable at https://react-redux.js.org/introduction/why-use-react-redux .

Also left some comments on that `connect` API doc draft.

**Comment by timdorr** [comment]
One thing I'd love to see in the docs: Testing! Patterns and guidelines to help folks out. I see a lot of confusion around this (especially with Hooks). We have a small section in the Redux core docs that should be migrated here: https://redux.js.org/recipes/writing-tests#connected-components

**Comment by armstrongl** [comment]
It doesn't look like this issue has had much activity recently, but I'm interested in helping out. I'm a technical writer and I'm currently learning React/Redux. If there's any documentation that's still missing or needs review (that can be understood by a beginner), I'd absolutely love to help out!

**Comment by timdorr** [comment]
@armstrongl Testing!

**Comment by markerikson** [comment]
@armstrongl , @timdorr : yep, docs on testing would be great.  Also, any of the other unchecked items from the issue post should still be up for grabs.  I'd be happy to help mentor you through working on any docs topics you'd like to contribute!

**Comment by markerikson** [comment]
Couple more things that we ought to cover in detail now that hooks are out:

- Hooks-specific details in the perf page we need to write
- Tradeoffs of using `connect` vs using hooks

Really, we probably ought to have a hooks page in the "Usage Guide" section.  Not entirely sure what it should cover, but throwing that out there.

**Comment by armstrongl** [comment]
@markerikson I would love to collaborate with you on these! I'm rather new to programming and, needless to say React/Redux, but I'm picking it up quickly and I'll be able to use my noobness in the docs to 'lower the barrier to entry' so to speak

**Comment by markerikson** [comment]
@armstrongl : yeah, having an outside eye on things would be great (both from the "tech writer" perspective and a "new to this stuff" perspective).

Tell you what.  Perhaps a good first task might be to do a review of the docs content we have right now.  That would both help us fix any outstanding issues, and help make sure you're familiar with most of the concepts around React-Redux.  That sound good?

**Comment by armstrongl** [comment]
@markerikson : sounds good! :) 

I'll post an update here once I feel I'm ready.

**Comment by wgao19** [comment]
Hey @markerikson, I've been working on Flow recently and I have [some notes](https://gist.github.com/wgao19/dfec19df95230949c610d2cadf4a40d5) on annotating connected components after upgrading Flow past 0.85. I think this version may have caused many confusions. Many people are unable to upgrade their codebases past 0.85 and / or may not know how to annotate connected components properly.

Would you like me to develop the notes into a doc for React Redux? And I'd like to hear your suggestions on the writing as well.

**Comment by markerikson** [comment]
Yeah, we really need a "Static Types" docs page in general.  I've added that to the list.

@jessidhia , would you have any time / interest in helping fill out the TS half of a "Static Types" docs page?

**Comment by ThiefMaster** [comment]
Docs on how to test components using the `useSelector` hook would be nice. E.g. whether people should wrap in Provider with a mocked store, or mock useSelector, etc.

**Comment by markerikson** [comment]
Repeating myself, I _really_ want a "Static Types" docs page.  It should definitely cover use with TS, and also Flow if possible.

I specifically want it to cover using the `ConnectedProps<T>` type described in these links:

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/37300

https://gist.github.com/christianchown/05e084c78ec216070a5a2b80f0534d4b

https://dev.to/voronar/practical-typescript-react--redux-4enh

https://github.com/DefinitelyTyped/DefinitelyTyped/issues/31227

**Comment by JNaftali** [comment]
I'm kinda interested in taking a swing at this. What other things would you want to be covered?

My initial thoughts:

- Typing connect and components
  - using ConnectedProps
  - using type arguments (`connect<StateProps, DispatchProps, OwnProps>`)
- Typing the store
  - Typical example
  - code splitting/lazy loading (`type RootState = SliceOne & SliceTwo`)

Should selectors be part of it even though reselect is it's own package?

**Comment by markerikson** [comment]
@JNaftali : great, thanks!

Yeah, I think the primary points would be around correctly typing `connect` and the hooks.  I'd say some mentions of typing selectors would be appropriate, as that relates to both `mapState` and `useSelector`.

I don't think we need to cover the store here, unless there's some very specific aspects with code splitting that really relate to React-Redux specifically.

**Comment by JNaftali** [comment]
Sounds good. I'll see about getting a rough draft done ASAP so you (or anyone else) can give feedback

**Comment by JNaftali** [comment]
Oh, one more thing. You say you want the page to be "Static Types": does that include Flow? My background is all Typescript but I imagine their usage is similar enough that I could figure it out?

**Comment by markerikson** [comment]
Let's start with TypeScript for now, as that's more widely used.

I know @wgao19 uses Flow, and she'd previously worked on a Flow-related page that I don't think we ever merged in.  She might be able to adapt some of that material for this page.

**Comment by wgao19** [comment]
Yah @markerikson I do have an open PR for this. Anything we shall add / edit to shape that up for what you have in mind about this section?

**Comment by markerikson** [comment]
Mmm... I'd have to go back and review what you'd written before.

I think my general observation was that the existing PR seems a bit too specific to "migrating past 0.85" or something, vs just "this is the right way to declare types for `connect`.  Can you rework it along those lines, and maybe we pick it up from there?

**Comment by renjithspace** [comment]
Hi guys, the documentation is very hard to learn for beginners. I just get started to read the doc and I facing difficulty to understand it. The `Get started` and `Tutorial` look the same as `API reference`. I think here needs a `Step by step` documentation approach like react did. And don't need to give complex logic in examples. I mean don't need to discuss the Todo app in examples. I think just a counter with increment and decrement is enough.

**Comment by markerikson** [comment]
@renjithspace : I'm not sure what problems you're trying to describe there.  What do you mean by "Get started and tutorial look the same as the API reference"?

**Comment by renjithspace** [comment]
@markerikson I mean even on the introductory part of the documentation trying to explain about the concept of the API. It’s like written the documentation from API perspective, not from the users (developers) perspective. It’s just my feedback that I felt when I started to read the documentation. I think the same is happening to the beginners like me.

Here is a tweet regarding the same https://twitter.com/dan_abramov/status/1039570011986321408?s=20 If I didn’t misunderstand what he was meant.

​
I watched a few tutorials available on youtube and I can see they explaining the basics in a simple way. That why I felt the same. Here is a reference: https://youtu.be/CVpUuw9XSjY

Thanks

**Comment by rajanlagah** [comment]
Hey @markerikson 
I have used react-redux in almost all of my react apps. I have good experience in REACT.  And I would love to collaborate with you on these! 
Will you please onboard me ?

**Comment by Gabri3l** [comment]
> Hey @markerikson
> I have used react-redux in almost all of my react apps. I have good experience in REACT. And I would love to collaborate with you on these!
> Will you please onboard me ?

You can start by reading the [Contributing](https://github.com/reduxjs/react-redux/blob/master/CONTRIBUTING.md) guidelines. I would recommend trying to focus on an open issue and have specific questions.
